server/automute_channel.go

```diff
@@ -0,0 +1,94 @@
+	"github.com/mattermost/mattermost/server/public/plugin"
+
+	"github.com/pkg/errors"
+)
+
+func (p *Plugin) UserHasJoinedChannel(c *plugin.Context, channelMember *model.ChannelMember, actor *model.User) {
+	_, _ = p.updateAutomutingOnUserJoinedChannel(c, channelMember.UserId, channelMember.ChannelId)
+}
+
+func (p *Plugin) updateAutomutingOnUserJoinedChannel(c *plugin.Context, userID string, channelID string) (bool, error) {
+	if automuteEnabled := p.getAutomuteIsEnabledForUser(userID); !automuteEnabled {
+		return false, nil
+	}
+
+	if canAutomute, err := p.isChannelLinked(channelID); err != nil {
+		p.API.LogError(
+			"Unable to check if channel is linked to update automuting when a user has joined the channel",
+			"UserID", userID,
+			"ChannelID", channelID,
+			"Error", err.Error(),
+		)
+		return false, errors.Wrap(err, "Unable to update automuting when a user has joined a channel")
+	} else if !canAutomute {
+		// Only automute channels that are linked
+		return false, nil
+	}
+
+	err := p.setChannelMembersAutomuted([]*model.ChannelMemberIdentifier{{UserId: userID, ChannelId: channelID}}, true)
+	return err == nil, err
+}
+
+func (p *Plugin) ChannelHasBeenCreated(c *plugin.Context, channel *model.Channel) {
+	_ = p.updateAutomutingOnChannelCreated(channel)
+}
+
+func (p *Plugin) updateAutomutingOnChannelCreated(channel *model.Channel) error {
+	if !channel.IsGroupOrDirect() {
+		// Assume that newly created channels can never be linked by the time this is called
+		return nil
+	}
+
+	return p.updateAutomutingForChannelMembers(channel.Id, true)
+}
+
+func (p *Plugin) updateAutomutingOnChannelLinked(channelID string) error {
+	// This simply mutes the channel for all users with automuting enabled, regardless of their settings before. It
+	// doesn't pay attention to if the user manually muted the channel beforehand.
+	return p.updateAutomutingForChannelMembers(channelID, true)
+}
+
+func (p *Plugin) updateAutomutingOnChannelUnlinked(channelID string) error {
+	// This simply unmutes the channel for all users with automuting enabled, regardless of their settings before. It
+	// doesn't pay attention to if the user manually muted the channel beforehand to keep it muted.
+	return p.updateAutomutingForChannelMembers(channelID, false)
+}
+
+func (p *Plugin) updateAutomutingForChannelMembers(channelID string, enableAutomute bool) error {
+	var membersToUpdate []*model.ChannelMemberIdentifier
+
+	page := 0
+	perPage := 200
+	for {
+		members, appErr := p.API.GetChannelMembers(channelID, page, 200)
+		if appErr != nil {
+			return errors.Wrap(appErr, fmt.Sprintf("Unable to get all members of channel %s to update automuting", channelID))
+		}
+
+		for _, member := range members {
+			if p.getAutomuteIsEnabledForUser(member.UserId) {
+				membersToUpdate = append(membersToUpdate, &model.ChannelMemberIdentifier{ChannelId: channelID, UserId: member.UserId})
+			}
+		}
+
+		if len(members) < perPage {
+			break
+		}
```

Comment:
nit: if `len(members) == perPage`, but we're at the end of the list, we'll go through the loop once more anyway and check this a second time. There's nothing wrong with that, it just make this terminating condition slightly unclear. I suppose we could always iterate until `len(members) == 0`. -1/5, just thinking out loud.

Reply:
Yeah, this is the same flaw in our paging API that we run into in frontend code as well where the caller has to guess if there's more results. It's a bit less expensive here than with an HTTP request, but it would be tremendously better if the APIs were explicit if there were more results or not. That's easy to implement because we just query the DB for `perPage + 1` results on the server side and use that extra result to know whether to set `hasMore` to true, but it would require changes to every paging API we have